In [1]:
import pandas as pd

In [2]:
users=pd.read_csv('takehome_users.csv',encoding='ISO-8859-1')

In [3]:
users

object_id        creation_time               name  \
0              1  2014-04-22 03:53:30     Clausen August   
1              2  2013-11-15 03:45:04      Poole Matthew   
2              3  2013-03-19 23:14:52  Bottrill Mitchell   
3              4  2013-05-21 08:09:28    Clausen Nicklas   
4              5  2013-01-17 10:14:20          Raw Grace   
...          ...                  ...                ...   
11995      11996  2013-09-06 06:14:15       Meier Sophia   
11996      11997  2013-01-10 18:28:37      Fisher Amelie   
11997      11998  2014-04-27 12:45:16        Haynes Jake   
11998      11999  2012-05-31 11:55:59       Faber Annett   
11999      12000  2014-01-26 08:57:12         Lima Thaís   

                            email     creation_source  \
0        AugustCClausen@yahoo.com        GUEST_INVITE   
1          MatthewPoole@gustr.com          ORG_INVITE   
2      MitchellBottrill@gustr.com          ORG_INVITE   
3       NicklasSClausen@yahoo.com        GUEST_INVITE   
4              GraceRaw@yahoo.com        GUEST_INVITE   
...                           ...                 ...   
11995       SophiaMeier@gustr.com          ORG_INVITE   
11996      AmelieFisher@gmail.com  SIGNUP_GOOGLE_AUTH   
11997         JakeHaynes@cuvox.de        GUEST_INVITE   
11998          mhaerzxp@iuxiw.com   PERSONAL_PROJECTS   
11999   ThaisMeloLima@hotmail.com              SIGNUP   

       last_session_creation_time  opted_in_to_mailing_list  \
0                    1.398139e+09                         1   
1                    1.396238e+09                         0   
2                    1.363735e+09                         0   
3                    1.369210e+09                         0   
4                    1.358850e+09                         0   
...                           ...                       ...   
11995                1.378448e+09                         0   
11996                1.358275e+09                         0   
11997                1.398603e+09                         1   
11998                1.338638e+09                         0   
11999                1.390727e+09                         0   

       enabled_for_marketing_drip  org_id  invited_by_user_id  
0                               0      11             10803.0  
1                               0       1               316.0  
2                               0      94              1525.0  
3                               0       1              5151.0  
4                               0     193              5240.0  
...                           ...     ...                 ...  
11995                           0      89              8263.0  
11996                           0     200                 NaN  
11997                           1      83              8074.0  
11998                           0       6                 NaN  
11999                           1       0                 NaN  

[12000 rows x 10 columns]

In [4]:
engagement=pd.read_csv('takehome_user_engagement.csv')

In [5]:
engagement

time_stamp  user_id  visited
0       2014-04-22 03:53:30        1        1
1       2013-11-15 03:45:04        2        1
2       2013-11-29 03:45:04        2        1
3       2013-12-09 03:45:04        2        1
4       2013-12-25 03:45:04        2        1
...                     ...      ...      ...
207912  2013-09-06 06:14:15    11996        1
207913  2013-01-15 18:28:37    11997        1
207914  2014-04-27 12:45:16    11998        1
207915  2012-06-02 11:55:59    11999        1
207916  2014-01-26 08:57:12    12000        1

[207917 rows x 3 columns]

In [6]:
engagement.visited.value_counts()

1    207917
Name: visited, dtype: int64

In [7]:
engagement.sort_values(by=['user_id','time_stamp'],inplace=True)

In [8]:
engagement

time_stamp  user_id  visited
0       2014-04-22 03:53:30        1        1
1       2013-11-15 03:45:04        2        1
2       2013-11-29 03:45:04        2        1
3       2013-12-09 03:45:04        2        1
4       2013-12-25 03:45:04        2        1
...                     ...      ...      ...
207912  2013-09-06 06:14:15    11996        1
207913  2013-01-15 18:28:37    11997        1
207914  2014-04-27 12:45:16    11998        1
207915  2012-06-02 11:55:59    11999        1
207916  2014-01-26 08:57:12    12000        1

[207917 rows x 3 columns]

In [9]:
engagement.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 207917 entries, 0 to 207916
Data columns (total 3 columns):
 #   Column      Non-Null Count   Dtype 
---  ------      --------------   ----- 
 0   time_stamp  207917 non-null  object
 1   user_id     207917 non-null  int64 
 2   visited     207917 non-null  int64 
dtypes: int64(2), object(1)
memory usage: 6.3+ MB


In [10]:
engagement['time_stamp'] = pd.to_datetime(engagement['time_stamp'])

In [11]:
engagement_shifted = engagement.shift(2)

In [12]:
engagement_shifted

time_stamp  user_id  visited
0                      NaT      NaN      NaN
1                      NaT      NaN      NaN
2      2014-04-22 03:53:30      1.0      1.0
3      2013-11-15 03:45:04      2.0      1.0
4      2013-11-29 03:45:04      2.0      1.0
...                    ...      ...      ...
207912 2014-04-11 19:39:56  11992.0      1.0
207913 2013-10-22 13:33:17  11995.0      1.0
207914 2013-09-06 06:14:15  11996.0      1.0
207915 2013-01-15 18:28:37  11997.0      1.0
207916 2014-04-27 12:45:16  11998.0      1.0

[207917 rows x 3 columns]

In [13]:
engagement_shifted.columns = ['time_stamp_shifted','user_id_shifted','visited_shifted']

In [14]:
engagement_shifted

time_stamp_shifted  user_id_shifted  visited_shifted
0                      NaT              NaN              NaN
1                      NaT              NaN              NaN
2      2014-04-22 03:53:30              1.0              1.0
3      2013-11-15 03:45:04              2.0              1.0
4      2013-11-29 03:45:04              2.0              1.0
...                    ...              ...              ...
207912 2014-04-11 19:39:56          11992.0              1.0
207913 2013-10-22 13:33:17          11995.0              1.0
207914 2013-09-06 06:14:15          11996.0              1.0
207915 2013-01-15 18:28:37          11997.0              1.0
207916 2014-04-27 12:45:16          11998.0              1.0

[207917 rows x 3 columns]

In [15]:
engagement = pd.concat([engagement, engagement_shifted], axis=1)

In [16]:
engagement

time_stamp  user_id  visited  time_stamp_shifted  \
0      2014-04-22 03:53:30        1        1                 NaT   
1      2013-11-15 03:45:04        2        1                 NaT   
2      2013-11-29 03:45:04        2        1 2014-04-22 03:53:30   
3      2013-12-09 03:45:04        2        1 2013-11-15 03:45:04   
4      2013-12-25 03:45:04        2        1 2013-11-29 03:45:04   
...                    ...      ...      ...                 ...   
207912 2013-09-06 06:14:15    11996        1 2014-04-11 19:39:56   
207913 2013-01-15 18:28:37    11997        1 2013-10-22 13:33:17   
207914 2014-04-27 12:45:16    11998        1 2013-09-06 06:14:15   
207915 2012-06-02 11:55:59    11999        1 2013-01-15 18:28:37   
207916 2014-01-26 08:57:12    12000        1 2014-04-27 12:45:16   

        user_id_shifted  visited_shifted  
0                   NaN              NaN  
1                   NaN              NaN  
2                   1.0              1.0  
3                   2.0              1.0  
4                   2.0              1.0  
...                 ...              ...  
207912          11992.0              1.0  
207913          11995.0              1.0  
207914          11996.0              1.0  
207915          11997.0              1.0  
207916          11998.0              1.0  

[207917 rows x 6 columns]

In [17]:
engagement.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 207917 entries, 0 to 207916
Data columns (total 6 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   time_stamp          207917 non-null  datetime64[ns]
 1   user_id             207917 non-null  int64         
 2   visited             207917 non-null  int64         
 3   time_stamp_shifted  207915 non-null  datetime64[ns]
 4   user_id_shifted     207915 non-null  float64       
 5   visited_shifted     207915 non-null  float64       
dtypes: datetime64[ns](2), float64(2), int64(2)
memory usage: 11.1 MB


In [18]:
engagement = engagement.iloc[2:,]
engagement = engagement.astype({'user_id_shifted':'int','visited_shifted':'int'})

In [19]:
engagement

time_stamp  user_id  visited  time_stamp_shifted  \
2      2013-11-29 03:45:04        2        1 2014-04-22 03:53:30   
3      2013-12-09 03:45:04        2        1 2013-11-15 03:45:04   
4      2013-12-25 03:45:04        2        1 2013-11-29 03:45:04   
5      2013-12-31 03:45:04        2        1 2013-12-09 03:45:04   
6      2014-01-08 03:45:04        2        1 2013-12-25 03:45:04   
...                    ...      ...      ...                 ...   
207912 2013-09-06 06:14:15    11996        1 2014-04-11 19:39:56   
207913 2013-01-15 18:28:37    11997        1 2013-10-22 13:33:17   
207914 2014-04-27 12:45:16    11998        1 2013-09-06 06:14:15   
207915 2012-06-02 11:55:59    11999        1 2013-01-15 18:28:37   
207916 2014-01-26 08:57:12    12000        1 2014-04-27 12:45:16   

        user_id_shifted  visited_shifted  
2                     1                1  
3                     2                1  
4                     2                1  
5                     2                1  
6                     2                1  
...                 ...              ...  
207912            11992                1  
207913            11995                1  
207914            11996                1  
207915            11997                1  
207916            11998                1  

[207915 rows x 6 columns]

In [20]:
engagement.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 207915 entries, 2 to 207916
Data columns (total 6 columns):
 #   Column              Non-Null Count   Dtype         
---  ------              --------------   -----         
 0   time_stamp          207915 non-null  datetime64[ns]
 1   user_id             207915 non-null  int64         
 2   visited             207915 non-null  int64         
 3   time_stamp_shifted  207915 non-null  datetime64[ns]
 4   user_id_shifted     207915 non-null  int64         
 5   visited_shifted     207915 non-null  int64         
dtypes: datetime64[ns](2), int64(4)
memory usage: 11.1 MB


In [21]:
engagement['time_dif'] = engagement['time_stamp']-engagement['time_stamp_shifted']

In [22]:
engagement = engagement[engagement['user_id']==engagement['user_id_shifted']]

In [23]:
engagement = engagement[engagement['time_dif']<='7 days']

In [24]:
active_users = set(engagement['user_id'])

In [25]:
users['adopted_user'] = [ int(x in active_users ) for x in users['object_id']]

In [26]:
features_at_sign_up=users[['creation_source','opted_in_to_mailing_list','enabled_for_marketing_drip','org_id','invited_by_user_id','adopted_user']]

In [27]:
features_at_sign_up

creation_source  opted_in_to_mailing_list  \
0            GUEST_INVITE                         1   
1              ORG_INVITE                         0   
2              ORG_INVITE                         0   
3            GUEST_INVITE                         0   
4            GUEST_INVITE                         0   
...                   ...                       ...   
11995          ORG_INVITE                         0   
11996  SIGNUP_GOOGLE_AUTH                         0   
11997        GUEST_INVITE                         1   
11998   PERSONAL_PROJECTS                         0   
11999              SIGNUP                         0   

       enabled_for_marketing_drip  org_id  invited_by_user_id  adopted_user  
0                               0      11             10803.0             0  
1                               0       1               316.0             1  
2                               0      94              1525.0             0  
3                               0       1              5151.0             0  
4                               0     193              5240.0             0  
...                           ...     ...                 ...           ...  
11995                           0      89              8263.0             0  
11996                           0     200                 NaN             0  
11997                           1      83              8074.0             0  
11998                           0       6                 NaN             0  
11999                           1       0                 NaN             0  

[12000 rows x 6 columns]

In [28]:
features_at_sign_up.org_id.value_counts()

0      319
1      233
2      201
3      168
4      159
      ... 
396      9
397      8
400      8
386      7
416      2
Name: org_id, Length: 417, dtype: int64

with 417 seperate organizations that would be a lot of fields to test over

In [29]:
features_at_sign_up['invited_by_adopted_user'] = [0 if pd.isna(x) else features_at_sign_up.iloc[int(x)-1,5] for x in features_at_sign_up['invited_by_user_id']]

/var/folders/qc/g6yf7hvd019dnqj11v94qpbw0000gn/T/ipykernel_15698/3626450820.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  features_at_sign_up['invited_by_adopted_user'] = [0 if pd.isna(x) else features_at_sign_up.iloc[int(x)-1,5] for x in features_at_sign_up['invited_by_user_id']]


In [30]:
features_at_sign_up = features_at_sign_up.drop(columns=['invited_by_user_id'])

In [31]:
features_at_sign_up

creation_source  opted_in_to_mailing_list  \
0            GUEST_INVITE                         1   
1              ORG_INVITE                         0   
2              ORG_INVITE                         0   
3            GUEST_INVITE                         0   
4            GUEST_INVITE                         0   
...                   ...                       ...   
11995          ORG_INVITE                         0   
11996  SIGNUP_GOOGLE_AUTH                         0   
11997        GUEST_INVITE                         1   
11998   PERSONAL_PROJECTS                         0   
11999              SIGNUP                         0   

       enabled_for_marketing_drip  org_id  adopted_user  \
0                               0      11             0   
1                               0       1             1   
2                               0      94             0   
3                               0       1             0   
4                               0     193             0   
...                           ...     ...           ...   
11995                           0      89             0   
11996                           0     200             0   
11997                           1      83             0   
11998                           0       6             0   
11999                           1       0             0   

       invited_by_adopted_user  
0                            0  
1                            0  
2                            1  
3                            1  
4                            1  
...                        ...  
11995                        0  
11996                        0  
11997                        0  
11998                        0  
11999                        0  

[12000 rows x 6 columns]

In [32]:
features_at_sign_up.groupby(['org_id']).mean().sort_values(by=['adopted_user'])

opted_in_to_mailing_list  enabled_for_marketing_drip  adopted_user  \
org_id                                                                       
416                     0.000000                    0.500000      0.000000   
151                     0.235294                    0.176471      0.000000   
183                     0.272727                    0.090909      0.000000   
237                     0.222222                    0.333333      0.000000   
242                     0.291667                    0.250000      0.000000   
...                          ...                         ...           ...   
415                     0.125000                    0.000000      0.375000   
399                     0.307692                    0.153846      0.384615   
270                     0.214286                    0.142857      0.428571   
235                     0.615385                    0.307692      0.461538   
387                     0.333333                    0.250000      0.583333   

        invited_by_adopted_user  
org_id                           
416                    0.000000  
151                    0.000000  
183                    0.000000  
237                    0.000000  
242                    0.000000  
...                         ...  
415                    0.312500  
399                    0.461538  
270                    0.357143  
235                    0.230769  
387                    0.333333  

[417 rows x 4 columns]

What orgaization a user is from strongly determines probaility of adopted user status. Some orgs have no adopted users while others have over 30%.

In [33]:
features_at_sign_up.groupby(['invited_by_adopted_user']).mean().sort_values(by=['adopted_user'])

opted_in_to_mailing_list  enabled_for_marketing_drip  \
invited_by_adopted_user                                                         
0                                        0.248482                    0.149457   
1                                        0.259259                    0.148148   

                             org_id  adopted_user  
invited_by_adopted_user                            
0                        139.969906      0.130407  
1                        160.231041      0.210758

A user has a 13% chance of being an adopted user if they are not invited by an adopted user. They have a 21% of being an adopted user if they are invited by an adopted user

In [34]:
features_at_sign_up.groupby(['creation_source']).mean().sort_values(by=['adopted_user'])

opted_in_to_mailing_list  enabled_for_marketing_drip  \
creation_source                                                            
PERSONAL_PROJECTS                   0.242539                    0.143534   
ORG_INVITE                          0.253409                    0.151387   
SIGNUP                              0.251557                    0.152851   
GUEST_INVITE                        0.250116                    0.147480   
SIGNUP_GOOGLE_AUTH                  0.244043                    0.149458   

                        org_id  adopted_user  invited_by_adopted_user  
creation_source                                                        
PERSONAL_PROJECTS   144.294173      0.081478                 0.000000  
ORG_INVITE          138.940762      0.134932                 0.176775  
SIGNUP              146.248203      0.144705                 0.000000  
GUEST_INVITE        141.343504      0.170596                 0.176607  
SIGNUP_GOOGLE_AUTH  141.523466      0.172563                 0.000000

How one signed up has an influence on use. Users invited by a guest to an organization or signed up with a google authorization have a 17% chance of being adopted users. People who signedup to another person's personal project have an 8% chance.

In [35]:
features_at_sign_up.groupby(['opted_in_to_mailing_list']).mean().sort_values(by=['adopted_user'])

enabled_for_marketing_drip      org_id  \
opted_in_to_mailing_list                                           
0                                           0.049967  141.639129   
1                                           0.448230  142.622912   

                          adopted_user  invited_by_adopted_user  
opted_in_to_mailing_list                                         
0                             0.136243                 0.093271  
1                             0.143287                 0.098196

There is only a slight difference between those who did and did not sign-up for the mailing list.

In [36]:
features_at_sign_up.groupby(['enabled_for_marketing_drip']).mean().sort_values(by=['adopted_user'])

opted_in_to_mailing_list      org_id  \
enabled_for_marketing_drip                                         
0                                           0.161834  141.402527   
1                                           0.748884  144.630580   

                            adopted_user  invited_by_adopted_user  
enabled_for_marketing_drip                                         
0                               0.137049                 0.094632  
1                               0.143415                 0.093750

likewise, marketing-drip-enabled users not substantially more likely to become adopted users.

Conclusions: The best indicators for if a user will become an adopted user from initial log-in information are 1)the org that they are a member of and 2) if they were invited by an adopted user.